src/library/scala/collection/IndexedSeq.scala

```diff
@@ -25,6 +25,28 @@ trait IndexedSeq[+A] extends Seq[A]
   override protected[this] def stringPrefix: String = "IndexedSeq"
 
   override def iterableFactory: SeqFactory[IndexedSeq] = IndexedSeq
+
+  override def foldLeft[B](z: B)(f: (B, A) => B): B = {
+    var b = z
+    var i = 0
+    while (i < length) {
+      val a = apply(i)
```

Comment:
`apply` is known to be a not-too-slow operation in `IndexedSeq`, but it is _not_ known to be faster than an iterator.  Therefore, the optimal form is just to use the iterator.  (Isn't that how it's implemented already in the superclass?)

Reply:
>  but it is _not_ known to be faster than an iterator. Therefore, the optimal form is just to use the iterator.

I am not sure that is what you meant to say - it doesn't really make sense. "`apply` is not _known_ to be faster" doesn't not mean that is  "_known_ to be slower". Its just not known. It need to compare in a benchmark, and for a few different list types (which will have different behaviour for `apply` and `Iterator`), different sizes (to see if that is an effect) etc. 

Then when we do know what is faster or slower we use the appropriate optimisations maybe for different places in the hierarchy when we learnt what is best. 

The iterator approach will also create an `Iterator` and IMHO its always worth running the jmh with --prof gc to see the memory load in the `gc.alloc.rate.norm`, so we get an idea of the CPU and memory effect or both alternatives

Reply:
What I meant was that being `IndexedSeq` is not supposed to mean that indexing is faster than iteration, just that it's a decently fast operation.  I actually know which way it is for a bunch of collections.  For ArrayBuffer, indexing is faster.  For Vector, iterating is faster.  Etc..  But if all you know is `IndexedSeq`, you can't assume indexing is faster.  In fact, it might be _quite a bit_ slower (e.g. Vector), in contrast to iteration which is usually only a little slower when it is slower.